src/transport/MessageHeader.cpp

```diff
@@ -158,60 +165,63 @@ CHIP_ERROR PacketHeader::Decode(const uint8_t * data, size_t size, size_t * deco
 
     if (mFlags.value & Header::Flags::kDestinationNodeIdPresent)
     {
-        VerifyOrExit(size >= kNodeIdSizeBytes, err = CHIP_ERROR_INVALID_ARGUMENT);
+        VerifyOrExit(size >= (p - data) + kNodeIdSizeBytes, err = CHIP_ERROR_INVALID_ARGUMENT);
         mDestinationNodeId.SetValue(LittleEndian::Read64(p));
-        size -= kNodeIdSizeBytes;
     }
     else
     {
         mDestinationNodeId.ClearValue();
     }
 
-    VerifyOrExit(size >= sizeof(uint16_t) * 2, err = CHIP_ERROR_INVALID_ARGUMENT);
+    static_assert(CHAR_BIT == 8, "We're assuming sizeof(uint16_t) is 2 bytes");
+    VerifyOrExit(size >= (p - data) + sizeof(uint16_t) * 2, err = CHIP_ERROR_INVALID_ARGUMENT);
     mEncryptionKeyID = LittleEndian::Read16(p);
     mPayloadLength   = LittleEndian::Read16(p);
 
-    *decode_len = p - data;
+    VerifyOrExit(p - data == EncodeSizeBytes(), err = CHIP_ERROR_INTERNAL);
+    static_assert(std::is_same<decltype(EncodeSizeBytes()), uint16_t>::value, "Unexpected EncodeSizeBytes() return type");
+    *decode_len = static_cast<uint16_t>(p - data);
 
 exit:
 
     return err;
 }
 
-CHIP_ERROR PayloadHeader::Decode(Header::Flags flags, const uint8_t * data, size_t size, size_t * decode_len)
+CHIP_ERROR PayloadHeader::Decode(Header::Flags flags, const uint8_t * const data, size_t size, uint16_t * decode_len)
 {
     CHIP_ERROR err    = CHIP_NO_ERROR;
     const uint8_t * p = data;
 
-    // TODO: this seems odd - know the size before decoding
-    // likely because header cross-referencing
-    VerifyOrExit(size >= EncodeSizeBytes(), err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(size >= kEncryptedHeaderSizeBytes, err = CHIP_ERROR_INVALID_ARGUMENT);
```

Comment:
This may not be accurate: encrypted header size is variable because of the inclusion of vendor id.

Reply:
This is why I added the `size` check in the `flags.value & Header::Flags::kVendorIdPresent` case.  This initial check ensures that we have enough buffer to read the always-present parts of the encrypted header, and the added check makes sure we can read a vendor id if asked to do it.

The new code definitely looks more correct to me than the old code, which could read off the end of the buffer afaict.  Again, I'm happy to do this some other way if we prefer; this was just what I came up with.